### PR TITLE
azurerm_hdinsight_kafka_cluster: support rest proxy

### LINKED ISSUE #8053: Support Kafka-Rest proxy provisioning on HDinsight Kafka Cluster
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Can we support Kafka rest proxy provisioning using terraform script.

```
"kafkaRestProperties": {
  "clientGroupInfo": {
    "groupName": "{AAD-Groupname}",
    "groupId": "{groupId}"
  }
},
```

### New or Affected Resource(s)

* azurerm_hdinsight_kafka_cluster

### PR DESCRIPTION
**Pull request by magodo**
This PR add supports for kafka rest proxy. To use kafka rest proxy, user will needs to specify both the rest prox setting and also a "kafka management node". In return, we will need to expose the endpoint for the kafka rest proxy.

Fixes #8053

### CODE REVIEW

**Comment by patelprakashp** [comment]
@magodo - Thank you very much. Much appreciated speedy response here. @tombuildsstuff @kosinsky - Can you please review changes and merge to main. We would like to use this feature soon.Thank you everyone.

**Comment by patelprakashp** [comment]
Can someone please push the changes to main branch @magodo @katbyte @kosinsky

**Comment by patelprakashp** [comment]
@kosinsky @magodo @katbyte @caius - Can someone please review and merge the changes to master branch.

**Comment by tombuildsstuff** [comment]
@magodo can you rebase this PR?

**Comment by patelprakashp** [comment]
@tombuildsstuff - Is there possibilities to merge this to master branch one day and available as a feature? I see this is being not included into weekly release since last 3 weeks , despite sign off from author and reviewer. cc: @magodo @katbyte @kosinsky.

**Comment by magodo** [comment]
@tombuildsstuff I merged the master into this branch as that is easier to resolve the conflict.

**Comment by patelprakashp** [comment]
@tombuildsstuff - can you please place milestone for next release ?

**Comment by patelprakashp** [comment]
CC : @tombuildsstuff - Can you please tag this for next release I see its not included v2.25.0 you please mark for v2.26.0. We would like to use terraform as possible. Also if there is no short term plan please let us know , We'll move to ARM template.

File: website/docs/r/hdinsight_kafka_cluster.html.markdown
Comment (tombuildsstuff) [inline website/docs/r/hdinsight_kafka_cluster.html.markdown:327]: The term "AAD" is great if you're familiar with Azure, but Azure Active Directory is clearer, so it's worth using that instead here

File: azurerm/internal/services/hdinsight/hdinsight.go
Comment (tombuildsstuff) [inline azurerm/internal/services/hdinsight/hdinsight.go:759]: fwiw you could make this:

```suggestion
		Required: required,
		Optional: !required,
		MaxItems: 1,
```

**Comment by tombuildsstuff** [comment]
@patelprakashp respectfully, would you mind refraining from @-ing people asking when this’ll be merged?

Whilst I appreciate you’d like to see this feature merged, unfortunately each notification is more noise for maintainers, which ultimately means that things take longer to get reviewed since we have more noise to filter out - [we instead ask that you use reactions (👍🏻) on the Issue/PR](https://www.terraform.io/docs/providers/azurerm/index.html#features-and-bug-requests) so that we can gauge interest. For the moment I’ve marked the “when will this be merged” comments as off-topic.

**Review by tombuildsstuff — CHANGES_REQUESTED** [review]
hey @magodo

Thanks for this PR / rebasing this.

Taking a look through this is looking good - if we can fix up the comments here then we should be able to take another look 👍

Thanks!

File: azurerm/internal/services/hdinsight/hdinsight_kafka_cluster_resource.go
Comment (magodo) [reply azurerm/internal/services/hdinsight/hdinsight_kafka_cluster_resource.go:341]: This should do no harm (and I refer to the same way which has been done for the [`EdgeNodeDef`](https://github.com/terraform-providers/terraform-provider-azurerm/blob/cffbabbcdf733487b7e926ca0e0fc5d932b2a12c/azurerm/internal/services/hdinsight/hdinsight_ml_services_cluster_resource.go#L172)). Because when you do expand, it will further check whether user has specify the corresponding schema; when you do flatten, it will further find the corresponding property by name (in this case `kafkamanagementnode`) in the response body. In both cases, if the kafka management node is not enabled, it shall do nothing.

File: azurerm/internal/services/hdinsight/hdinsight.go
Comment (magodo) [reply azurerm/internal/services/hdinsight/hdinsight.go:759]: Thank you for the hint!

File: azurerm/internal/services/hdinsight/hdinsight_kafka_cluster_resource.go
Comment (magodo) [reply azurerm/internal/services/hdinsight/hdinsight_kafka_cluster_resource.go:131]: API will asks for this property, otherwise it will complain lack of the group name. I think the capacity to look up the group name based on the group id is out of the scope of azurerm provider, it is handled in azuread provider I assume.

I agree it seems unnecessary from the user's perspective, but I don't know how we can improve this given current API behavior.

File: azurerm/internal/services/hdinsight/hdinsight_kafka_cluster_resource.go
Comment (magodo) [reply azurerm/internal/services/hdinsight/hdinsight_kafka_cluster_resource.go:416]: The ID here is a UUID, so it is not feasible to get the group name from it.

File: azurerm/internal/services/hdinsight/hdinsight_kafka_cluster_resource.go
Comment (tombuildsstuff) [reply azurerm/internal/services/hdinsight/hdinsight_kafka_cluster_resource.go:133]: I mean we should file an API bug here, since providing both seems unnecessary - in the interim we can look that up using the Azure AD API's here?

File: azurerm/internal/services/hdinsight/hdinsight_kafka_cluster_resource.go
Comment (tombuildsstuff) [reply azurerm/internal/services/hdinsight/hdinsight_kafka_cluster_resource.go:417]: as above, we can look this up in the Azure AD API

**Comment by magodo** [comment]
@tombuildsstuff 

I have made the change that use aad client to get the group name instead of asking user to provide. Please take another look!

## Test Result

```bash
💤 make testacc TEST=./azurerm/internal/services/hdinsight/tests TESTARGS='-run TestAccAzureRMHDInsightKafkaCluster_restProxy'
                                                                                                                                                              
==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
==> Checking that acceptance test packages are used...                                                                                                        
TF_ACC=1 go test ./azurerm/internal/services/hdinsight/tests -v -run TestAccAzureRMHDInsightKafkaCluster_restProxy -timeout 180m -ldflags="-X=github.com/terra
form-providers/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccAzureRMHDInsightKafkaCluster_restProxy                                                                                                       
=== PAUSE TestAccAzureRMHDInsightKafkaCluster_restProxy                                                                                                       
=== CONT  TestAccAzureRMHDInsightKafkaCluster_restProxy
--- PASS: TestAccAzureRMHDInsightKafkaCluster_restProxy (1502.83s)                                                                                            
PASS   
ok      github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/hdinsight/tests     1502.859s
```

**Comment by patelprakashp** [comment]
@magodo @tombuildsstuff - anything I can help here?

**Comment by katbyte** [comment]
@magodo - could we get this rebased? should be good to merge once thats done

**Comment by magodo** [comment]
@katbyte I've resolved the conflicts while the test failed, will update once it passes...

**Review by katbyte — APPROVED** [review]
Thanks @magodo - this LGTM 👍

**Pull request merged**
